src/protocols/secure_channel/PASESession.cpp

```diff
@@ -523,52 +599,49 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
-    VerifyOrExit(buf_len == sizeof(encryptionKeyId) + kMAX_Point_Length, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
 
-    err = mSpake2p.BeginVerifier(nullptr, 0, nullptr, 0, mPASEVerifier.mW0, kSpake2p_WS_Length, mPoint, sizeof(mPoint));
-    SuccessOrExit(err);
+    const uint8_t * X;
+    size_t X_len = 0;
+
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-    encryptionKeyId = chip::Encoding::LittleEndian::Read16(buf);
-    msg->ConsumeHead(sizeof(encryptionKeyId));
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == 1, err = CHIP_ERROR_INVALID_TLV_TAG);
+    X_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(X));
+    SuccessOrExit(
+        err = mSpake2p.BeginVerifier(nullptr, 0, nullptr, 0, mPASEVerifier.mW0, kSpake2p_WS_Length, mPoint, sizeof(mPoint)));
 
     // Pass Pa to check abort condition.
-    err = mSpake2p.ComputeRoundOne(msg->Start(), msg->DataLength(), Y, &Y_len);
-    SuccessOrExit(err);
+    SuccessOrExit(err = mSpake2p.ComputeRoundOne(X, X_len, Y, &Y_len));
+    SuccessOrExit(err = mSpake2p.ComputeRoundTwo(X, X_len, verifier, &verifier_len));
 
-    ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", encryptionKeyId);
-    SetPeerKeyId(encryptionKeyId);
+    {
+        size_t data_len = EstimateTLVStructOverhead(Y_len + verifier_len, 2);
 
-    err = mSpake2p.ComputeRoundTwo(msg->Start(), msg->DataLength(), verifier, &verifier_len);
-    SuccessOrExit(err);
+        System::PacketBufferHandle msg2 = System::PacketBufferHandle::New(data_len);
+        VerifyOrExit(!msg2.IsNull(), err = CHIP_ERROR_NO_MEMORY);
 
-    // Make sure our addition doesn't overflow.
-    VerifyOrExit(UINTMAX_MAX - verifier_len >= Y_len, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
-    VerifyOrExit(CanCastTo<uint16_t>(Y_len + verifier_len), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
-    data_len = static_cast<uint16_t>(sizeof(encryptionKeyId) + Y_len + verifier_len);
+        System::PacketBufferTLVWriter tlvWriter;
+        tlvWriter.Init(std::move(msg2));
 
-    {
-        Encoding::LittleEndian::PacketBufferWriter bbuf(System::PacketBufferHandle::New(data_len));
-        VerifyOrExit(!bbuf.IsNull(), err = CHIP_ERROR_NO_MEMORY);
-        bbuf.Put16(GetLocalKeyId());
-        bbuf.Put(&Y[0], Y_len);
-        bbuf.Put(verifier, verifier_len);
-        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+        SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(1), ByteSpan(Y, Y_len)));
```

Comment:
Here again--> `Y` and `pB` are related:  `then computes Y=y*P+w0*N, and transmits pB=Y to A`.

The "verifier" in the next field is actually `cB`.

In Spake2+:

```
                   A                           B

                   |         (Preamble)        |
                   |<- - - - - - - - - - - - ->|
                   |                           |
                   |       (setup protocol)    |
      (compute pA) |             pA            |
                   |-------------------------->|
                   |             pB            | (compute pB)
                   |<--------------------------|
                   |                           |
                   |       (derive secrets)    | (compute cB)
                   |             cB            |
                   |<--------------------------|
      (compute cA) |             cA            |
                   |-------------------------->|
```

We use Pake1 for `pA` and Pake1 for `pB || cB` in one physical message to cover the 2 logical messages.

I think it would be worthwhile to keep the names X and Y, and use the `kPake1_pB`, `kPake2_pB` and `kPake2_cB` for the tag constant names to make clear we are filling these named fields per spec.